### PR TITLE
feat: display permission level in /query

### LINKED ISSUE #624: `/query` enhancement: show permissions
Currently query only shows payment related information. We should display all the info we have on the user, which at least includes the permissions for them to be able to set labels on the repository.

### PR DESCRIPTION
**Pull request by EtherealGlow**
Resolves #624

### CODE REVIEW

**Review by whilefoo — CHANGES_REQUESTED** [review]
it should show permissions from `access` database table. Also don't forget about QA

**Comment by EtherealGlow** [comment]
@whilefoo i am getting a whole charades of weird errors, mostly `Error logging to Supabase: undefined`

**Comment by EtherealGlow** [comment]
the bot has worked previously before idk whats happening now

**Comment by EtherealGlow** [comment]
should i use markup table for displaying all 4 accesses and their levels like
`
| access type | access level |
| ----------- | ------------ |
| multiplier  | true         |
| priority    | true         |
| time        | true         |
| price       | true         |
`

**Comment by EtherealGlow** [comment]
for now i ma gonna this

**Review by whilefoo — CHANGES_REQUESTED** [review]
could you make a function that retrieve all access levels at the same time to avoid multiple database calls

**Comment by pavlovcik** [comment]
Please make sure that CI passes @EtherealGlow

**Comment by EtherealGlow** [comment]
Done

**Comment by EtherealGlow** [comment]
Uhh maybe not 😬

File: src/adapters/supabase/helpers/client.ts [hunk 1]
```diff
@@ -340,6 +347,20 @@ export const getAccessLevel = async (username: string, repository: string, label
...
+export const getAllAccessLevels = async (username: string, repository: string): Promise<false | accessLevels> => {
+  const logger = getLogger();
+  const { supabase } = getAdapters();
+
+  const { data } = await supabase.from("access").select("*").eq("user_name", username).eq("repository", repository).single();
+
+  if (!data) {
+    logger.info(`Access not found on the database`);
+    // no access
+    return false;
```
Comment (whilefoo) [inline src/adapters/supabase/helpers/client.ts:359]:
```suggestion
    return null;
```
I think `null` is more appropriate

File: src/adapters/supabase/helpers/client.ts [hunk 2]
```diff
@@ -5,6 +5,13 @@ import { Database } from "../types";
 import { InsertPermit, Permit } from "../../../helpers";
 import { BigNumber, BigNumberish } from "ethers";
 
+interface accessLevels {
```
Comment (whilefoo) [inline src/adapters/supabase/helpers/client.ts:8]:
```suggestion
interface AccessLevels {
```
be mindful about naming conventions

File: src/handlers/comment/handlers/query.ts [hunk 3]
```diff
@@ -22,13 +22,23 @@ export const query = async (body: string) => {
   const regex = /^\/query\s+@([\w-]+)\s*$/;
   const matches = body.match(regex);
   const user = matches?.[1];
+  const repo = payload.repository;
 
   if (user) {
+    const data = await getAllAccessLevels(user, repo.full_name);
+    if (data === false) {
```
Comment (whilefoo) [inline src/handlers/comment/handlers/query.ts:29]:
```suggestion
    if (!data) {
```

File: src/handlers/comment/handlers/query.ts [hunk 4]
```diff
@@ -22,13 +22,23 @@ export const query = async (body: string) => {
...
     if (!walletInfo?.address) {
       return `Error retrieving multiplier and wallet address for @${user}`;
     } else {
-      return `@${user}'s wallet address is ${walletInfo?.address} and  multiplier is ${walletInfo?.multiplier}`;
+      return `@${user}'s wallet address is ${walletInfo?.address}, multiplier is ${walletInfo?.multiplier} and access levels are
+| access type | access level |
+| multiplier  |  ${data["multiplier"]}      |
+| priority    |  ${data["priority"]}        |
+| time        |  ${data["time"]}            |
+| price       |  ${data["price"]}           |`;
```
Comment (whilefoo) [inline src/handlers/comment/handlers/query.ts:41]:
```suggestion
| price       |  ${data.price}           |`;
```
same for others

**Review by whilefoo — CHANGES_REQUESTED** [review]
Please post QA

File: src/adapters/supabase/helpers/client.ts (hunk 1 above)
Comment (0xcodercrane) [reply src/adapters/supabase/helpers/client.ts:359]: it can be either `null` or `undefined`. Literally it doesn't make sense to return `boolean` type

**Comment by EtherealGlow** [comment]
@whilefoo I can't post the qa because the bot ain't working for me

**Review by whilefoo — CHANGES_REQUESTED** [review]
https://github.com/ubiquity-whilefoo/ubiquibot/issues/82#issuecomment-1708055079 please fix the table, it's not displaying correctly

**Comment by EtherealGlow** [comment]
ok wait

**Review by whilefoo — CHANGES_REQUESTED** [review]
I think it's unnecessary to use an external library for a simple table

**Comment by EtherealGlow** [comment]
@whilefoo might work now

**Review by whilefoo — CHANGES_REQUESTED** [review]
```
| Command | Description |
| --- | --- |
| git status | List all new or modified files |
| git diff | Show file differences that haven't been staged |
```
this is how you make a table

**Comment by EtherealGlow** [comment]
I am dumb 💀

File: src/adapters/supabase/helpers/client.ts [hunk 5]
```diff
@@ -340,6 +347,20 @@ export const getAccessLevel = async (username: string, repository: string, label
...
+  const { supabase } = getAdapters();
+
+  const { data } = await supabase.from("access").select("*").eq("user_name", username).eq("repository", repository).single();
+
+  if (!data) {
+    logger.info(`Access not found on the database`);
+    // no access
+    return null;
+  }
+  return { multiplier: data["multiplier_access"], time: data["time_access"], priority: data["priority_access"], price: data["price_access"] };
```
Comment (whilefoo) [inline src/adapters/supabase/helpers/client.ts:361]:
```suggestion
  return { multiplier: data.multiplier_access, time: data.time_access, priority: data.priority_access, price: data.price_access };
```

**Review by whilefoo — CHANGES_REQUESTED** [review]

**Review by whilefoo — APPROVED** [review]
QA: https://github.com/ubiquity-whilefoo/ubiquibot/issues/89

**Pull request merged**